### ISSUE TITLE
 Expose SILO functionality via a REST interface

### ISSUE THREAD

**Issue by fengelniederhammer** [GenSpectrum/LAPIS-SILO#2]
Implement a REST interface that one can send SILO queries via a HTTP request to.

@chaoran-chen @Taepper How should we integrate the "update data" step? Should we implement another endpoint that triggers loading new data? Or is there another way to access the data used by a running SILO instance?

**Comment by Taepper** [comment]
Maybe a trigger via the same REST interface is a good idea, if LAPIS triggers it?
The data itself rather over the file system?
I would say SILO expects a version number for the data and then all the .fasta and .tsv files we mentioned last time.
I will implement the respective versioned update procedure.

**Comment by fengelniederhammer** [comment]
If LAPIS is supposed to trigger the update, we need to secure it via access keys or similar. We don't want anyone to be able to trigger updates via a public API.

My first intuition would be to rather let SILO compute the version number. Maybe hash the metadata file, if that is suitable?

How would the update look like? Something like that maybe?
POST `/triggerUpdate` with URLS to files
-> SILO responds with a version number and starts processing the files in the background.

POST `/loadUpdate` with version number
-> SILO responds with either
*  "update not yet ready" because the pipeline is still running
* "ok" and loads the serialized data that the pipeline produced

**Comment by chaoran-chen** [comment]
For the version number, I suggest using a timestamp. It could be of the time when new data were found or when we finished the processing of the new data.

**Comment by Taepper** [comment]
After more thinking, I think it should rather not be an API call.
On start of SILO, I would just load the data via file system or remote file location.
I think using the time-stamp for versioning is fine.
I can then keep track of that time-stamp to decide whether to load the last state or start the pipeline anew.

So I would think we would need to decide, how the files (aligned.fasta, metadata.tsv, ...) will be made available.
How to keep track of the timestamp. (i) use file time-stamp? (ii) additional file for the timestamp (e.g. "version.info")

I would be happy to implement anything from the point when I have a handle to the directory where I can find the set of files

**Comment by fengelniederhammer** [comment]
So how does a running LAPIS instance get updates then? It would check for a new version on a regular schedule in a configured location?

How do you trigger the update as a maintainer?

**Comment by Taepper** [comment]
For the update procedure it does not matter whether SILO is restarted or updates while running. So can LAPIS not start a new SILO server with the updated files?

**Comment by fengelniederhammer** [comment]
What would "LAPIS starts a new SILO" look like on a technical level?

What I imagine the solution to look like is: LAPIS has a URL configured where it should send the SILO queries to. I expect to let LAPIS and SILO communicate only via this web interface to have a good separation between the artifacts.

I can't yet think of a good way to implement such a "restart SILO" command.

**Comment by chaoran-chen** [comment]
Are we talking about "new flat files are available" (i.e., pre-SILO-preprocessing) or "new serialized state is available" (post-SILO-preprocessing)? Or both?

For the SILO-preprocessing, I think that it should be independent of the always-running API server (they could both be started from the same binary and on the server but it would be a separate process). I can then imagine that the preprocessing can be executed with a command like `./silo preprocess --input ~/input-data-1676987734 --output ~/new-serialized-state-1676987734`. It is up to the maintainer to run the SILO-preprocessing whenever they like.

Once the new serialized state is ready, it needs to be made available to the SILO-API server. Here, I could indeed imagine a file system-based solution. They would be a directory with sub-directories where each sub-directory is a serialized state. The sub-directories are named by the version numbers. E.g.,

```
~/data/
    |---- 1676986139
    |---- 1676987734
```

The SILO-API server will then regularly check the folder. When it detects a state with a higher version number, it loads it.

**Comment by fengelniederhammer** [comment]
Ok, that sounds like a reasonable solution for now. It sounds tedious if you have to login yourself into the server, but we can automate this more at a later point.

I wonder how we should implement it for this ticket. Should we split it into two parts?
* Create a web interface that cannot be updated at runtime (this issue)
* Implement the "look for updates" of the API (separate issue)

**Comment by chaoran-chen** [comment]
The suggested split sounds good to me!

**Comment by fengelniederhammer** [comment]
I turned #7 into an epic and made the "look for updates" a sub task of it.
